GoMeet/src/main/java/com/google/sps/servlets/GoogleCalendarTimesServlet.java

```diff
@@ -0,0 +1,185 @@
+          ErrorMessages.INVALID_ARG_FORMAT);
+      return;
+    }
+
+    List<TimePeriod> times = autoProposeTimes(response, guestList, 
+        periodStart, periodEnd, totalDurationMs);
+    if (times == null) {
+      return; // The error message would have been sent by autoProposeTimes.
+    }
+
+    // Build JSON response.
+    response.setContentType("application/json");
+    response.getWriter().println(timePeriodListToJson(times));
+  }
+
+  /**
+   * Converts a List<TimePeriod> to a JSON String that represents a List of 
+   * Strings. Each String is the datetime string of the START time of each
+   * TimePeriod (i.e. the meeting time proposed will be at the beginning of each 
+   * free TimePeriod in times)
+   * @param times The List of TimePeriods representing the free times proposed
+   * by AutoProposeTimes.
+   * @return The String in JSON notation of the resultant list of possible
+   * meeting times. 
+   */
+  private String timePeriodListToJson(List<TimePeriod> times) {
+    // Return times that are the start times of each TimePeriod in times
+    List<String> timesStr = new ArrayList<String>();
+    for (int i = 0; i < times.size(); i++) {
+      String current = times.get(i).getStart().toStringRfc3339();
+      timesStr.add(current);
+    }
+    Gson gson = new Gson();
+    return gson.toJson(timesStr);
+  }
+  
+  /**
+   * Instantiates the AutoProposeTimes class and calls the proposeTimes()
+   * function, returning the result from the proposeTimes() call.
+   * NOTE: This code encapsulated in a function to make it easier to
+   * mock out the calls to AutoProposeTimes in the tests for the servlet.
+   * @param response The response from the servlet, to send error messages on
+   * failures.
+   * @param guestList The list of guests' calendar IDs
+   * @param periodStart The start of the period to generate times for.
+   * @param periodEnd The end of the period
+   * @param totalDurationMs Duration of the meeting to be scheduled, in milliseconds.
+   * @return The List of TimePeriod as returned by a successful call to proposeTimes, 
+   * or null if the call returned errors.
+   * @see AutoProposeTimes class
+   */
+  public List<TimePeriod> autoProposeTimes(HttpServletResponse response, 
+      ArrayList<String> guestList, Date periodStart, 
+      Date periodEnd, int totalDurationMs) throws IOException {
+    
+    // Ensure that AutoProposeTime is working in the same timezone as this servlet.
+    AutoProposeTimes.setTimezone(TIMEZONE); 
+    
+    AutoProposeTimes proposer;
+    List<TimePeriod> times;
+    
+    try {
+      Calendar service = new Calendar.Builder(GoogleNetHttpTransport.newTrustedTransport(),
+        GsonFactory.getDefaultInstance(), 
+        null)
+        .setApplicationName("GoMeet")
+        .build();
+      proposer = new AutoProposeTimes(service, API_KEY, guestList, 
+          periodStart, periodEnd, totalDurationMs);
+      times = proposer.proposeTimes();
```

Comment:
Just for conversation, in hindsight, it might not have been necessary to create an object with internal state to do this since the usage model is instantiate, execute, destroy. I wonder if it would have sufficed just used a singe static function to take all the dependencies and arguments instead.

Reply:
Hmm yeah I think that can be something to refactor in the future.